JavaScript based REST interface client

s-ludwig opened the pull request.

Geod24 commented:
Any ETA for when you want this ? I'll review by the end of this week.
Though at first sight it looks like a there's a fair amount of duplication ?

s-ludwig commented:
I have actually started to work on the duplication issue (which is already present even without this addition), but the current `RestInterface` struct intended for this is still not ready for CT generation, because it currently requires the `RestInterfaceSettings` for a lot of things. I'll work on that later.

I'll probably create a 0.7.24 branch during the next couple of days and start merging your remaining changes in `vibe.web.rest`. Then I'd do a larger refactoring, moving things out of the `vibe.web.rest` module and switching over to use `RestInterface` there, too.

s-ludwig commented:
Cleaned up the commit history for review.

Geod24 commented on source/vibe/web/internal/rest/common.d:
```diff
@@ -0,0 +1,407 @@
+	header,      // req.header[]
+	attributed,  // @before
+	internal     // req.params[]
+}
```
IIUC, internal is for path attribute, isn't it ? Maybe that'll be a more descriptive name. Also body parameters are not meant to stay JSON forever. On that topic: any way to check the server setting ? It's not a big deal, but if  someone disable `parseQueryString` for example, there's no way to detect it.

s-ludwig replied:
It's for path attributes or any other attributes that are passed using `req.params`. Since there are valid uses outside of path placeholders, I decided against using that as the name. With the "JSON" comment I just wanted to describe the current state, since I'm not sure yet how to best handle multiple formats in terms of the internal architecture.

Currently there is no way to check the server settings. Really the only possibility that I see would be to throw an exception from inside `HTTPServerRequest` if the `.json` field is accessed without the `parseJson` setting. Another, probably better, path would be to switch the whole JSON body parsing to happen lazily and optionally strongly typed.

Geod24 commented on source/vibe/web/internal/rest/jsclient.d:
```diff
@@ -0,0 +1,106 @@
+module vibe.web.internal.rest.jsclient;
```
It could really use some documentation / documented unittest :)

s-ludwig replied:
I'll add an example in `vibe.web.rest` and the missing module header, as well as a doc string for `generateInterface` here.

Geod24 commented on examples/rest-js/dub.json:
```diff
@@ -0,0 +1,8 @@
+		"vibe-d": { "path": "../../" }
+	},
+	"versions": ["VibeDefaultMain"]
+}
```
Nitpick: Missing newline.

As per the example that follow, there's no comment so it's a bit hard on newcomers to see how useful it is.

s-ludwig replied:
Will fix.

Geod24 commented on source/vibe/web/rest.d:
```diff
@@ -248,6 +179,47 @@ unittest
 /**
+	Returns a HTTP handler delegate that serves a JavaScript REST client.
+*/
+HTTPServerRequestDelegate serveRestJSClient(TImpl)(RestInterfaceSettings settings = null)
```
I would only allow interfaces for `Timpl` here.

s-ludwig replied:
Hm, you are right, I copied this from the server.

Geod24 commented on source/vibe/web/rest.d:
```diff
@@ -248,6 +179,47 @@ unittest
+/**
+	Generates JavaScript code to access a REST interface from the browser.
+*/
+void generateRestJSClient(TImpl, R)(ref R output, RestInterfaceSettings settings = null)
```
Interface contraint + `isOutputRange!R`

s-ludwig replied:
Will fix.

Geod24 commented on source/vibe/web/rest.d:
```diff
@@ -516,7 +488,7 @@ unittest
 	/** The public URL below which the REST interface is registered.
 	*/
-	URL baseURL;
+	URL baseURL = URL("http://api.example.com/");
```
Why ?

Geod24 commented on source/vibe/web/internal/rest/common.d:
```diff
@@ -0,0 +1,407 @@
+						case Header: pi.kind = ParameterKind.header; break;
+						case Query: pi.kind = ParameterKind.query; break;
+						case Body: pi.kind = ParameterKind.body_; break;
+					}
```
We could also kill `WebParamAttribute.Origin` and use `ParameterKind` instead.

Geod24 commented:
Hum, another P.R. blocked by 2.065 :(

Also, it looks like there's a discrepancy between the client and server with the refactoring:
```console
REST route: GET /root/get_check/:param/:param2 []
add route GET /root/get_check/:param/:param2
[...]
no route match: GET /root/getCheck/one/two
No response written for /root/getCheck/one/two
REST call: GET http://127.0.0.1:8000/root/getCheck/one/two -> 404, 404
```

For the rest, the diff is large, so I'll probably have to try it out on my machine to be sure, but I like where it's heading at. I'm not that familiar with JS, but the new functionality looks cool :+1:

s-ludwig commented:
Fixed the inconsistency (c5d3208) and 2.065 compilation error (ed739a9).

s-ludwig commented:
BTW, for xref purposes, `RestInterface` also makes things like #662 and #1070 an easy target.

s-ludwig commented:
Rebased on master. I've also fixed the handling of optional parameters. Commit ca3def1 needs some re-review, because during rebase it was just a single bit conflict with the write-to-headers change. Apart from that, everything except the default service URL comment and the possibly merge of `ParameterKind` and `WebParamAttribute.Origin` should be addressed.

I'll have a look at the latter now. Regarding the URL, I don't remember exactly anymore, but I think I got some internal assertions due to an empty service URL, which was quite unobvious. We could alternatively use something like `"http://localhost/"` or just add high-level assertions with a proper error message. The latter sounds like the better way to go...

s-ludwig commented:
@Geod24, did you see any remaining issues? Otherwise I'd rebase/squash and merge now.

Geod24 commented:
I'll do a last sweep this evening ;)

Geod24 commented:
Can't see anything.
All your project build out of the box with the new version ?

s-ludwig commented:
I haven't tested all of them yet, but I'll do that now.

s-ludwig commented:
Okay, when testing my projects I found one bug in the REST client (09ee6a27f9a57a4a9757dcc8aff073148be94d39), but none in the new code. Will merge once Travis goes green.

s-ludwig commented:
And thanks for reviewing this lengthy change set!

etcimon commented:
I would have loved to review this but got caught up in the busiest time of the past few years. Sorry!

ColdenCullen commented on source/vibe/web/internal/rest/jsclient.d:
```diff
@@ -0,0 +1,116 @@
+		// XHR setup
+		output.put("    var xhr = new XMLHttpRequest();\n");
+		output.formattedWrite("    xhr.open('%s', url, true);\n", route.method.to!string.toUpper);
+		static if (!is(RT == void)) {
```
Sorry for not catching this sooner, but is `RT` actually defined here? I think it's causing this block to execute regardless of return type.

s-ludwig replied:
Thanks, looks like that... `is()` and `__traits(compiles)` are really suboptimal constructs w.r.t. correctness. I'll add a test and fix this.

s-ludwig replied:
Err.. or rather just merge your PR and add a test ;)
